Cardano Shelley update - part 1

gabrielKerekes opened the pull request:
As promised we are splitting the Cardano Shelley update into (at least) 3 PRs. This is the first of them.

This PR contains minimal changes with which Trezor can work on the new Shelley fork. It enables sending funds from one Byron address to another. New addresses are not yet added neither are certificates and withdrawals. 

**What's changed:**
- Previous transactions (input sources) no longer need to be requested/received since Fee is included directly in the transaction.
- TTL parameter has been added to transaction - this is the block until which the transaction can be submited.
- Transaction format has changed a bit.
- The biggest changes are in the `sign_tx.py` file. The logic around fee calculation has been removed and it has also been refactored quite a lot - e.g. Transaction class has been removed since it now mostly contained only functions anyways.
- Support for CBOR NULL had to be added to support empty transaction metadata.
- Protocol magic is included in Byron addresses on testnets - this has always been a part of the spec, but it hasn't been needed so far.

**What can still change in the following days (regarding this PR):**
- We still don't have specs/information about witnesses for Byron inputs from IOHK. This will probably require changes to the way the witnesses are built for Byron inputs. There's also a todo there.
- Protocol magic in Byron addresses on Testnets - this functionality is drafted, but also needs to be confirmed by IOHK.
- Maybe some other minor changes.

[Here's a gist](https://gist.github.com/gabrielKerekes/b8b929614e11bea2ff8af6affe4caf0b) with a sample transaction input file for trezorctl (although tests also provide some samples).
[Here's a gist](https://gist.github.com/gabrielKerekes/85dd1c0bf1059e4c2627fc2ee719cb50) with a description of the Shelley transaction format (including certificates and withdrawals).
[Here's](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl) the official CDDL spec for Shelley.

**This hasn't been tested on the latest testnet yet, since Byron addresses are not yet fully supported. Support should be added in the following days. When this happens, we will test it and let you know of the status** (Some changes may still need to be made regarding this).

refi93 commented on common/protob/messages-cardano.proto:
```diff
@@ -50,22 +51,21 @@ message CardanoPublicKey {
     repeated CardanoTxOutputType outputs = 2;   // outputs to be used in transaction
-    optional uint32 transactions_count = 3;     // transactions count
-    optional uint32 protocol_magic = 5;         // network's protocol magic
+    optional uint32 protocol_magic = 3;         // network's protocol magic
```
IIRC reusing protobuf field ids is a no-no because if somebody sends a message following the old convention it may clash.

Was there a reason why you changed the id of the "protocol_magic" field?

refi93 replied:
http://itshark.xyz/posts/2018/7/11/Protocol_Buffers_basic_stuff_you_need_to_know -> `It is very important to remember that once some number is used in some message, it can never be changed or reused.`

gabrielKerekes replied:
I would've thought that it is only a problem if you want to keep some support for the old version. Since we're pretty much dropping support of the old version I thought it'd be OK. But sure, better safe than sorry - I'll fix it.

refi93 commented on common/protob/messages-cardano.proto:
```diff
@@ -50,22 +51,21 @@ message CardanoPublicKey {
         optional bytes prev_hash = 2;   // hash of previous transaction output to spend by this input
         optional uint32 prev_index = 3; // index of previous output to spend
-        optional uint32 type = 4;       // input type, defaults to 0
     }
```
given the suggestion above, I would just comment out but not remove this so future contributors do not reuse this id

refi93 commented on core/src/apps/cardano/__init__.py:
```diff
@@ -6,6 +6,9 @@
 CURVE = "ed25519"
 SEED_NAMESPACE = [HARDENED | 44, HARDENED | 1815]
 
+MAINNET_PROTOCOL_MAGIC = 764824073
```
is that really shelley mainnet protocol magic as well?

refi93 replied:
If not or we are not sure, I would rather remove this and retain only the testnet magic (42 apparently) - it's not a breaking thing and better to not mislead users into thinking they are interacting with the mainnet if they actually aren't

refi93 replied:
ok, agreed offline that this will be changed in later PRs anyway

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,42 @@
-ACCOUNT_PREFIX_DEPTH = const(2)
+ACCOUNT_PREFIX_DEPTH = const(3)
+
+LOVELACE_INVALID = 45_000_000_000 * 1_000_000
```
I think `LOVELACE_MAX_SUPPLY` would be more a accurate name

gabrielKerekes replied:
Yeah, I was deciding between the two - both are used in Ledger code. I'll rename it.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,42 @@
+def is_change(output: List[int], inputs: List[CardanoTxInputType]) -> bool:
     for input in inputs:
         inp = input.address_n
+        print(output[:ACCOUNT_PREFIX_DEPTH], inp[:ACCOUNT_PREFIX_DEPTH])
```
probably a forgotten debug statement

gabrielKerekes replied:
Yes! Thanks for catching it.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
+    def _build_inputs(self) -> list:
+        inputs_for_cbor = []
+        for input in self.inputs:
+            inputs_for_cbor.append([input.prev_hash, input.prev_index])
```
IMHO this deserves to be a tupple and there should be no conflict in encoding as I see that the cbor encoder does not have a rule for encoding tuples yet: https://github.com/vacuumlabs/trezor-firmware/blob/ca3dd2bfb9b02cfad8077f77e680740e6de3d9e0/core/src/apps/common/cbor.py#L70 so I would extend it to support them. 

This will allow you to put here a more specific return type which increases readability

refi93 replied:
same for the remaining `_build` functions

refi93 replied:
regarding the CBOR rule for tuples, that'sin hindsight probably too radical as this may complicate the decoding logic as there would not be 1:1 mapping as the same coding would be used for lists as for tuples. Still, you can return tuples from there and cast them to lists just before cbor-encoding

gabrielKerekes replied:
Yes, I though about that, I just didn't want to do `list(returned_value)` each time. I'll do it then.

gabrielKerekes replied:
Actually no. I don't think it'll work. I return lists from those functions. And those lists are then directly added to tx_body. So I would have iterate through them again. ... We'll discuss it offline.

gabrielKerekes replied:
In the end we decided to return lists of tuples from these functions. We then use a `_detupleize` function to convert those lists of tuples into lists of lists (as required by CBOR).

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
+                    self.keychain, output.address_n, self.protocol_magic
+                )
+                # set output address so it can later be shown to the user if needed
+                output.address = address
```
this kind of smells - one would not expect a "build" function to mutate the class properties. Isn't there a way to achieve the same without this hack?

gabrielKerekes replied:
This is needed just for the show_tx function. I wasn't able to come up with a better way. I can also add something like this to show_tx:
`        if not output.address:
            address, _ = derive_address_and_node(keychain, output.address_n)
        else:
            address = output.address

        await confirm_sending(ctx, output.amount, address)`
which would require sending in the keychain and deriving the address a second time. Or storing it in some property as it was before, but I wanted to simplify it.

janmazak replied:
What about making `show_tx` part of the `Transaction` class? The class has all the means to compute the addresses in whatever format is necessary, and it also knows the inputs and outputs to show.

gabrielKerekes replied:
To me it feels like showing the transaction is a different matter for which the Transaction class should not be responsible. Also, maybe we don't even the Transaction class anymore? (see comment below)

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
+    def _build_inputs(self) -> list:
```
`_prepare_inputs_for_serialization` or just `_prepare_inputs` would imho give to the reader enough info so you don't need to name the return variable `inputs_for_cbor` which kind of exposes in the function a lower level concept that is not really needed to understand it

gabrielKerekes replied:
How else can I name what I return? `prepared_inputs`? :shrug: It has no other use than to be used in CBOR, so I don't have a better name :(

refi93 replied:
Imho `result` would do - the type would be clear from the type definition if updated and the most valuable information for the reader at that point is that this variable is the one meant to be returned, so no need to complicate it, but it's a matter of taste, of course

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
-        self.fee = self.compute_fee(
-            self.input_coins_sum, self.outgoing_coins, self.change_coins
-        )
+        tx = cbor.encode([tx_body, witnesses, None])
```
I would name this serialised_tx

gabrielKerekes replied:
I'll rename it.

I need to build the `tx_body` beforehand in order to hash it for signing. So then only witnesses would remain to be added as you describe, but I am not sure it's worth it.

refi93 replied:
I see, ok nevermind

refi93 commented on core/src/apps/cardano/__init__.py:
```diff
@@ -6,6 +6,9 @@
 
+MAINNET_PROTOCOL_MAGIC = 764824073
+TESTNET_PROTOCOL_MAGIC = 1097911063
+
```
this imho deserves a `ProtocolMagic` enum: https://docs.python.org/3/library/enum.html - and you can save the use of "Mainnet", "testnet", ... strings and also helper dicts like `KNOWN_PROTOCOL_MAGICS` that way

gabrielKerekes replied:
The strings are needed in the show_tx function. So the map is a helper for that. I don't even need to know if the id is known. I only need to know if the one I received is a testnet one or not. ... We'll discuss it offline.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,42 @@
 # use and still allow to quickly brute-force the correct bip32 path
 MAX_CHANGE_ADDRESS_INDEX = const(1000000)
-ACCOUNT_PREFIX_DEPTH = const(2)
+ACCOUNT_PREFIX_DEPTH = const(3)
```
why did the account prefix depth increase? Did the derivation path format change?

gabrielKerekes replied:
It hasn't. This is also a forgotten debugging change....

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,42 @@
 
-# we consider addresses from the external chain as possible change addresses as well
-def is_change(output, inputs):
+# only addresses from the same account as the inputs are considered as change addresses
```
I also don't like much that it is not consistent with what the HD derivation scheme says are change addresses (change addresses have specific value of the 4th item in the derivation path). So we may want to keep this behavior (not showing such outputs), but let's at least call the concept differently (e.g. this function could be `hide_output`) --- how are the Trezor people supposed to figure out why we are not following the details of HD specs?

gabrielKerekes replied:
Yes, it would show all the outputs to the user. I think it's pretty harmless to show it to the user since he can simply check the address and this shouldn't even happen very often, right?

gabrielKerekes replied:
I took another look at this and I don't really understand what this function is supposed to achieve. The previous comment:
`we consider addresses from the external chain as possible change addresses as well` kind of describes what it does. It's not about accounts, it's about external chains (a path different from `m/44'/1815'` e.g. `m/44'/1111'` will be treated as change address).

However when I try to use a change output with path `m/44'/1111'/0'/0/1` I get a "Forbidden key path" error. Which would mean that an address from the external chain can never occur in the change outputs.

Any ideas?

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -40,62 +50,38 @@ def is_change(output, inputs):
+        await confirm_sending(ctx, output.amount, output.address)
 
+    if total_amount > LOVELACE_INVALID:
+        raise ValueError("Total transaction amount is out of range")
```
Isn't this a wire error, caused by user inputs?

gabrielKerekes replied:
Makes sense.

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -40,62 +50,38 @@ def is_change(output, inputs):
         )
 
+        if msg.fee > LOVELACE_INVALID:
+            raise ValueError("Fee is out of range")
```
wire error?

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
+    ) -> None:
         self.inputs = inputs
         self.outputs = outputs
         self.keychain = keychain
```
`keychain` is a different type of argument than the rest, so I would move it to the front of the list.

gabrielKerekes replied:
I agree, moved.

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
+        return outputs_for_cbor
 
-    def _build_witnesses(self, tx_aux_hash: str):
+    def _build_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes]]:
```
Now I'm not sure, but have we removed Byron witness computation? That does not sound right, we still need it for legacy utxos, don't we?

gabrielKerekes replied:
We haven't. We simply don't know for sure what it will be, so I have decided to leave it out until we do.

janmazak commented on python/src/trezorlib/cli/cardano.py:
```diff
@@ -58,11 +59,12 @@ def sign_tx(client, file, network):
     """Get Cardano address."""
     address_n = tools.parse_path(address)
-    return cardano.get_address(client, address_n, show_display)
+    return cardano.get_address(client, address_n, show_display, network)
```
Do we call
```
def get_address(
    client, address_n: List[int], protocol_magic: int, show_display=False
) -> messages.CardanoAddress:
```
?

If yes, I would not call the command line arg `--network` when it is passed as `protocol_magic`, it just adds to the confusion IOHK created.

gabrielKerekes replied:
Very true.

ppershing commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,42 @@
 from apps.common.paths import validate_path
 from apps.common.seed import remove_ed25519_prefix
 
+if False:
```
forgotten?

gabrielKerekes replied:
No, this is how types, which are only imported for the sake of type hinting, are imported throughout the project.

ppershing commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -40,62 +50,38 @@ def is_change(output, inputs):
+        total_amount += output.amount
+        await confirm_sending(ctx, output.amount, output.address)
 
+    if total_amount > LOVELACE_INVALID:
```
This is clearly wrong -- you skip changing total_amount for change addresses

gabrielKerekes replied:
True. I've moved the check to the `_build_outputs` function. Does it even make sense to have it?

ppershing commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
```
do we know what are these magic numbers? isn't it some cbor?

refi93 replied:
yes it's the CBOR prefix for a 32-byte byte sequence and this is probably a legacy from the way it was done in AdaLite: https://github.com/vacuumlabs/adalite/blob/bf73ee00cbd23f3648d80ba09e56873ca433479a/app/frontend/wallet/byron/cardano-wallet-secret-crypto-provider.ts#L89 - so it should be fine to do proper cbor encoding here

gabrielKerekes replied:
Yes it is. Previously it was also done this way although now that I think about it, it was probably because protocol magic was somehow magically encoded cbor, right? 

I've changed it to `cbor.encode(tx_aux_hash)`

ppershing commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
-            )
+    def serialise_tx(self) -> Tuple[bytes, bytes]:
```
I would split this into part that computes tx_body/tx_hash and the rest which does signing/witnesses

gabrielKerekes replied:
I've extracted `_build_tx_body()` and `_hash_tx_body()` but I've kept it in the `serialise_tx()` becuase I didn't like the idea of having to make two calls unnecessarily. Let me know if that's better.

As I was doing this it occured to me, that we maybe don't even need the Transaction class anymore. What do you think? Also @refi93 @janmazak ?

ppershing commented on tests/device_tests/test_msg_cardano_get_address.py:
```diff
@@ -26,24 +26,43 @@
-    "path,expected_address",
+    "path,protocol_magic,expected_address",
     [
+        # data from https://iancoleman.io/bip39/
```
afait data isn't from there

gabrielKerekes replied:
Honestly, I don't know. The comment was at the start of the test before, so I moved it up since it definitely isn't true for all the test cases now. Should I remove it since we aren't sure?

janmazak replied:
Remove it, won't be true for Shelley anyway.

gabrielKerekes commented on core/src/apps/cardano/address.py:
```diff
@@ -13,11 +14,19 @@ def _encode_address_raw(address_data_encoded):
+    # todo: GK - verify this is indeed correct (after IOHK confirmation/testnet support)
+    # protocol magic is included in Byron addresses only on testnets
+    if protocol_magic == MAINNET_PROTOCOL_MAGIC:
+        address_payload = None
+        address_attributes = {}
```
Yep. Done.

gabrielKerekes commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -40,62 +50,38 @@ def is_change(output, inputs):
-    raw_outputs: list,
 ) -> None:
+    total_amount = 0
     for index, output in enumerate(outputs):
```
Indeed.

gabrielKerekes commented:
Thanks everyone for the thorough review. I tried to implement all your suggestions.

One more thing you should all be aware of - yesterday on a call with @refi93 we decided to return lists of tuples instead of lists of lists in the `_build_x` functions. For that we also needed to implement `_detupleize` which converts those tuples to list. This was done in favor of readability and type hints. Let us know what you think :) (We weren' t so sure about what the correct location for the function was so we've put it in the `sign_tx.py` file outside the Transaction class - maybe an utils file would be suitable, but since the `_detupleize` function is pretty much local, we didn't like that)

gabrielKerekes commented:
I don't perticulalry like the name of `tx_body` in `CardanoSignedTx(tx_body=tx_body, tx_hash=tx_hash)` message . It's confusing with the real `tx_body` being something different.

Wouldn't it be a good idea to rename it to e.g. `serialised_tx`?

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,32 +4,37 @@
+def should_hide_output(output: List[int], inputs: List[CardanoTxInputType]) -> bool:
     for input in inputs:
         inp = input.address_n
+        print(output[:ACCOUNT_PREFIX_DEPTH])
```
forgotten print

gabrielKerekes replied:
Omg, and on the same spot. Sorry :disappointed: Removed.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -40,62 +44,42 @@ def is_change(output, inputs):
-
         transaction = Transaction(
-            msg.inputs, msg.outputs, keychain, msg.protocol_magic, input_coins_sum
+            keychain, msg.inputs, msg.outputs, msg.protocol_magic, msg.fee, msg.ttl,
```
as proposed offline, this class indeed seems redundant if it's used just to serialize the tx and we can just keep the `serialise_tx` call below. Btw, I noticed, other coins use the term `serialize`, not `serialise` so it may be worth staying consistent for better searchability

refi93 commented on common/protob/messages-cardano.proto:
```diff
@@ -77,26 +79,6 @@ message CardanoSignTx {
-
 /**
  * Response: Serialised signed cardano transaction
  * @end
```
as proposed offline, I agree the `tx_body` field should indeed be named `serialized_tx`  - that way it would be consistent with other coins and avoids confusion with "tx_body" as the structured payload of the transaction

gabrielKerekes commented:
Ok. I have removed the Transaction class. I have reordered the functions so they read top to bottom (it made the diff a bit mory messy, if it's too much of a problem, I can reorder them back). 

I have also renamed `tx_body` of `CardanoSignedTx` message to `serialized_tx`. 

I also added tests for the `_should_hide_output` function so that it is maybe clearer what it should do.

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,238 +4,181 @@
+                raise wire.ProcessError("Invalid output address!")
+
+        total_amount += amount
+        result.append((base58.decode(address), amount))
```
We should make sure using base58 is compatible with trezor connect --- I have no idea. (I would prefer to receive bytes, but we can change it later in another shelley PR).

gabrielKerekes replied:
It is compatible. It has worked the same way in Byron. Yes, we will be receiving bytes and it will be done in the second PR since it doesn't make sense here yet.

gabrielKerekes commented:
I have added a function `matches_with_protocol_magic` to `address.py`, which checks whether the output is meant for the given protocol magic. I've added test cases for this also.

I have left the `fixup` commits there for now, so it's easier to review only what has changed. I will rebase them later.

gabrielKerekes commented:
Closing and creating a new PR to Trezor repo.
